### DISCUSSION TITLE
I2C using two devices

### DISCUSSION THREAD

**Discussion by npant333** [zephyrproject-rtos/zephyr#42618] [Q&A]
Device 1
i2c1: arduino_i2c: i2c@40005400 {
   reg = < 0x40005400 0x400 >;
   ...
}

Device 2
i2c1: i2c@40004000 {
    reg = < 0x40004000 0x1000 >;
    ...
}

I am trying to transfer data form Device 1 to Device 2 using I2C (As described above)          
 
The following code (In Device 1)  fails

uint8_t buf[5] = { 0x43, 0x43, 0x43, 0x44, 0x45 };
ret = i2c_write(i2c_dev, buf, sizeof(buf), 40004000);

What is incorrect here?

**Comment by henrikbrixandersen** [comment]
I think we'll need a little more context than the small snippets of code, you provided, in order to give a meaningful answer.

**Reply by npant333** [reply]
Thanks for your feedback. Here are more details.

Before I did this, I tried the following experiment using Nucleo 429ZI (Using Zephyr Code) and Arduino UNO (Using Arduino IDE) and it was successful.
Connected the appropriate wires between F429ZI and UNO.

Code in UNO to read :
void setup() {
  Wire.begin(1);        // join i2c bus with address 1
  Serial.begin(9600);   // start serial for output
  Wire.onReceive(receiveEvent);
}

void loop() {
  Serial.print("In loop\n");
  delay(3000);
}

void receiveEvent(int howMany) {
 while (0 <Wire.available()) {
    char c = Wire.read();      /* receive byte as a character */
    Serial.print(c);           /* print the character */
  }
 Serial.println();             /* to newline */
}


Code in F429ZI for sending data to Slave address 1

const struct device *i2c_dev = DEVICE_DT_GET(DT_NODELABEL(i2c1));
uint32_t i2c_cfg;
i2c_cfg = I2C_MODE_MASTER | I2C_SPEED_SET(I2C_SPEED_STANDARD);
i2c_configure(i2c_dev, i2c_cfg);
    
uint8_t write_buf[7] = { 0x43, 0x43, 0x43, 0x44, 0x45 };

// Slave address = 1 (Setup in UNO code)
ret = i2c_write(i2c_dev, write_buf, 5, 1);
if (ret == 0)
{
    printk("Write data was fine\n");
}
else 
{
    printk("Write failed\n");
}


I can see "CCCDE" in UNO serial monitor

Now, I would like to communicate between a Nucleo F429ZI and Particle Xenon using I2C and both using Zephy code. Code given in my first post.

I looked at the dts and found the i2c1 address of the devices. I wrote the zephyr code for both F429ZI and Particle Xenon. But in both devices, the read and write failed. I replaced the UNO connection with Xenon (With F429ZI) using pins provided in  (https://docs.particle.io/datasheets/discontinued/xenon-datasheet/)

**Reply by ndrs-pst** [reply]
@npant333, it seems you're attempting to use one device as a controller (master) and another as a target (slave).
In this case, I suggest exploring the `drivers/i2c/target` directory for relevant configurations and examples. You might find these two Zephyr documentation links helpful for further understanding:

- [I2C Interface](https://docs.zephyrproject.org/3.5.0/hardware/peripherals/i2c.html) - for general information about the I2C interface in Zephyr.
- [I2C EEPROM Target](https://docs.zephyrproject.org/3.5.0/hardware/peripherals/i2c_eeprom_target.html) - for specific details about setting up an I2C EEPROM target.
